persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java

```diff
@@ -94,37 +114,179 @@ public void beforeChange(
 
   @Override
   public void onChange(InetAddress endpoint, ApplicationState state, VersionedValue value) {
+    if (state == ApplicationState.STATUS) {
+      return;
+    }
+
     EndpointState epState = Gossiper.instance.getEndpointStateForEndpoint(endpoint);
     if (epState == null || Gossiper.instance.isDeadState(epState)) {
       return;
     }
 
-    addStargateNode(endpoint, epState);
+    if (!isStargateNode(epState)) {
+      return;
+    }
+
+    if (!joinCluster(endpoint, epState)) {
+      applyState(endpoint, state, value, epState);
+    }
   }
 
   @Override
-  public void onAlive(InetAddress endpoint, EndpointState state) {}
+  public void onAlive(InetAddress endpoint, EndpointState state) {
+    if (!isStargateNode(state)) {
+      return;
+    }
+    notifyUp(endpoint);
+  }
 
   @Override
-  public void onDead(InetAddress endpoint, EndpointState state) {}
+  public void onDead(InetAddress endpoint, EndpointState state) {
+    if (!isStargateNode(state)) {
+      return;
+    }
+    notifyDown(endpoint);
+  }
 
   @Override
   public void onRemove(InetAddress endpoint) {
+    leaveCluster(endpoint);
+  }
+
+  @Override
+  public void onRestart(InetAddress endpoint, EndpointState state) {}
+
+  private boolean joinCluster(InetAddress endpoint, EndpointState state) {
+    if (!liveStargateNodes.add(endpoint)) {
+      return false;
+    }
+
+    updateTokens(endpoint);
+
+    for (Map.Entry<ApplicationState, VersionedValue> entry : state.states()) {
+      applyState(endpoint, entry.getKey(), entry.getValue(), state);
+    }
+
+    if (StorageService.instance.isRpcReady(endpoint)) {
+      notifyJoinCluster(endpoint);
+    } else {
+      endpointsPendingJoinedNotification.add(endpoint);
+    }
+
+    return true;
+  }
+
+  private void leaveCluster(InetAddress endpoint) {
     if (!liveStargateNodes.remove(endpoint)) {
       return;
     }
+    StargateSystemKeyspace.removeEndpoint(endpoint);
+    InetAddress nativeAddress = getNativeAddress(endpoint);
+    for (EventListener listener : listeners) {
+      listener.onLeaveCluster(nativeAddress, EventListener.NO_PORT);
+    }
+  }
 
-    for (IEndpointLifecycleSubscriber subscriber : subscribers) subscriber.onLeaveCluster(endpoint);
+  private void notifyJoinCluster(InetAddress endpoint) {
+    InetAddress nativeAddress = getNativeAddress(endpoint);
+    for (EventListener listener : listeners) {
+      listener.onJoinCluster(nativeAddress, EventListener.NO_PORT);
```

Comment:
maybe we could get rid of the extra `NO_PORT` case, by passing the `InetSocketAddress` everywhere?
We could construct the `InetSocketAddress` based on the `InetAddress` and `server.socket.getPort()` when the `InetAddress` is received (at the higher level), wdyt?